pkg/apis/kubedirector/v1beta1/kubedirectorcluster_types.go

```diff
@@ -25,11 +25,20 @@ import (
 // requested cluster roles, each of which will be implemented (by KubeDirector)
 // using a StatefulSet.
 type KubeDirectorClusterSpec struct {
-	AppID         string    `json:"app"`
-	AppCatalog    *string   `json:"appCatalog,omitempty"`
-	ServiceType   *string   `json:"serviceType,omitempty"`
-	Roles         []Role    `json:"roles"`
-	DefaultSecret *KDSecret `json:"defaultSecret,omitempty"`
+	AppID               string      `json:"app"`
+	AppCatalog          *string     `json:"appCatalog"`
+	ServiceType         *string     `json:"serviceType"`
+	Roles               []Role      `json:"roles"`
+	DefaultSecret       *KDSecret   `json:"defaultSecret"`
+	ConfigMetaGenerator int32       `json:"configMetaGenerator"`
+	Connections         Connections `json:"connections"`
+}
+
+// Connections specifies list of cluster objects and configmaps objects that has
+// be connected to the cluster.
+type Connections struct {
+	Clusters   []string `json:"clusters,omitempty"`
+	ConfigMaps []string `json:"configmaps,omitempty"`
```

Comment:
I think this is a case where generalizing-from-two is premature or incorrect. We handle these two kinds of "connections" somewhat differently, and there's no guarantee that there will be a third kind or how we might handle that. (I realize we sort of inherited this arrangement from EPIC, but I think we talked before about how it is probably not the right thing.)

I would recommend that we have these two lists NOT bundled together inside a Connections object. E.g. Connections (representing ONLY connected kdclusters) and ConfigMaps would be two separate and parallel properties with no wrapper object around them.

We should also have a serious think about whether either or both of these are the sort of thing that can be expressed per-role -- as is the case with secrets, labels, injected files, etc. etc. -- and not just at the top level spec.

Those decisions won't have much effect on the underlying logic of processing the configmaps and cluster connections, but we'll need to settle them before merging.

(Also of course there's the matter of making changes to the CRD. We have already put the stamp on v1beta1, so any changes we make here will need to be expressed in v1beta2. There's some stuff to figure out there in terms of how we author the CRD to express both v1beta1 and v1beta2, how we deal with existing resources, how we upgrade existing CRDs, and how we internally "promote" any v1beta1-styled resource inputs to store them as v1beta2. You and I and I think Koshik need to have a separate talk about that.)

Reply:
Hmm, I thought about this. Even if we put config maps outside of the connection we still support only specific k8s resources. This is just a way to express that some resources could be "connected" to cluster and we can then act/watch. We want to be able to communicate that certain resources could be "connected" to a cluster.  With or without the wrapper the data structures need to be adjusted accordingly.